### ISSUE TITLE
Styles: when lots of color presets exist, the experience gets laggy/jumpy to click between

### ISSUE THREAD

**Issue by annezazu** [WordPress/gutenberg#65686]
When using a theme, [like assembler
](https://github.com/Automattic/themes/tree/trunk/assembler), with lots of color presets, the sidebar gets very wonky when selecting between them, meaning the selection jumps around and is noticeably slow.

I can replicate this with WP 6.6.2 and Gutenberg 19.3:

https://github.com/user-attachments/assets/f118d585-8c34-480d-b7bd-b5d0f5891abe

**Comment by ramonjd** [comment]
Trying to think what would be a reasonable solution to this.

With `n` theme style variations containing colors, there's obviously a lot of computation going on in the background.

Here's the component for reference: [ColorVariations](https://github.com/WordPress/gutenberg/blob/0ed82aec5f97a6e5fc06d0654f645865d8484caa/packages/edit-site/src/components/global-styles/variations/variations-color.js#L17-L17)

A mitigation strategy might be to have a limit how many the sidebar renders, and have "Load more" or pagination functionality. 

This could solve both problems:

1. The lag
2. From a UI perspective, loading `n` palettes in the sidebar making things cluttered and crowded.

**Comment by ndiego** [comment]
> A mitigation strategy might be to have a limit how many the sidebar renders, and have "Load more" or pagination functionality.

This seems like a reasonable solution. Besides a theme like the Assembler, I cannot imagine many themes having so many color presets. Not sure what the max `n` would be. Perhaps 20?

**Comment by ramonjd** [comment]
20 would be a good start. If it's laggy in testing, then... 19. 😄

**Comment by richtabor** [comment]
At the same time, many themes will have this many variations. I'm not thinking this as high priority.

**Comment by andrewserong** [comment]
Oh, interesting that the clicking on a preset is resulting in snapping up to the top like that. There is a similar issue In Styles in the left hand browse mode, too (earlier issue https://github.com/WordPress/gutenberg/issues/61841), so I wonder if there's a common root cause 🤔

**Comment by andrewserong** [comment]
I suspect there are a few different places where performance needs to be optimised, but from a quick bit of hacking around, it seems that the context calculations for each of the previews winds up being recalculated every time a selection occurs. On the one hand this is "good" because it means the previews are always current _in case_ something changed, but in this situation we know that nothing did. I wonder if we can be smarter about when we re-use the existing context, versus recalculating? I haven't gotten very far in my digging, but here's a quick link to the `useMemo` I was looking at that appears to fire much more frequently than I'd expect (i.e. for every variation in the list, even though the variation itself hasn't changed when a user selects a different one):

https://github.com/WordPress/gutenberg/blob/e11e06a9717691e3cb1b8934d4cbea0c25a91a7d/packages/edit-site/src/components/global-styles/variations/variation.js#L37

**Comment by ndiego** [comment]
> At the same time, many themes will have this many variations. I'm not thinking this as high priority.

Given that 6.7 Beta 2 is releasing later today, do you think we should punt this to 6.8 @richtabor @andrewserong?

**Comment by andrewserong** [comment]
Thanks for the ping, Nick! The bug has been around for a long time (and I think was present in at least the previous release, too) so I've moved to to "punted to 6.8". Do feel free to move it back if anyone would like to — I'd love for us to find a fix for it, but I'll likely be a bit too short on time to dig in much further right now.

**Comment by t-hamano** [comment]
Let's move this issue to the 6.9 Editor Tasks board and see what's possible in 6.9.

**Comment by Mamaduka** [comment]
This will probably require larger refactoring.

Currently, Global Styles is shared/consumed using a large context object; every change will render all consumers (default React behavior), which makes it hard to apply some of the best practices.

**Comment by ramonjd** [comment]
> Currently, Global Styles is shared/consumed using a large context object; every change will render all consumers (default React behavior), which makes it hard to apply some of the best practices.

Thanks @Mamaduka 

I wanted to come back to this one day to try out [this idea](https://github.com/WordPress/gutenberg/issues/65686#issuecomment-2378154632) - not sure when I could do that right now 

I have not looked into this at all, but it sounds like we can't memoize our way out of it. Another thought that came to mind is breaking global styles up into smaller contexts, but now I'm just saying words...

**Comment by ramonjd** [comment]
> Another thought that came to mind is breaking global styles up into smaller contexts

Oh man, I saw this blast from the past just now:

- https://github.com/WordPress/gutenberg/issues/67534#issuecomment-3219821501

Thanks for cross-linking!

**Comment by t-hamano** [comment]
I do remember I was able to reproduce this problem in the past, but not now 🤔

There are no unnatural jumps or lags. I tested it with the [Assembler](https://github.com/Automattic/themes/tree/trunk/assembler) theme. The theme file is here:

[assembler.zip](https://github.com/user-attachments/files/21979272/assembler.zip)

Is anyone still able to reproduce this issue with the latest Gutenberg?

https://github.com/user-attachments/assets/22928be7-6842-4070-aeaf-735d71445d64

**Comment by andrewserong** [comment]
> Is anyone still able to reproduce this issue with the latest Gutenberg?

No, this is feeling pretty responsive to me in the latest Gutenberg while running the Assembler theme. I'll close this out now, but we can always re-open it if anyone finds it's still feeling laggy!

**Comment by Mamaduka** [comment]
Yes, it performs better than initially reported, but it still takes a couple of seconds for the change to be visible on the canvas.

I think responsiveness also depends on the complexity of a template. This could be the reason for the difference between @annezazu's and @t-hamano's screencast.

That being said, we can explore optimizing this screen further after https://github.com/WordPress/gutenberg/issues/67534.

## Screenshot

https://github.com/user-attachments/assets/3ad2bf4a-c829-4d84-a69a-0ffa01952d9f

**Comment by ramonjd** [comment]
Thanks for testing this again, folks!
